Issue 633 aws load balancer - closes #633 (#645)
bob-crutchley/courseware: topics/aws/modules/aws-load-balancer/README.md

[load-balancer-dashboard]: https://eu-west-1.console.aws.amazon.com/ec2/home?region=eu-west-1#LoadBalancers:sort=loadBalancerName

[ec2-search]: https://i.imgur.com/h9u5mP8.png

[add-tag]: https://i.imgur.com/h9B9hxz.png

[finding-security-group]: https://i.imgur.com/WJtAHZX.png

[adding-inbound-rule]: https://i.imgur.com/Uiud0Ys.png

[create-load-balancer-button]: https://i.imgur.com/opWv9I8.png

[create-application-load-balancer]: https://i.imgur.com/r7eHmye.png

[load-balancer-configuration]: https://i.imgur.com/LAJHd3v.png

[load-balancer-routing-config]: https://i.imgur.com/0uXbf5Y.png

[load-balancer-add-instances]: https://i.imgur.com/BWrqTUR.png

[load-balancer-DNS]: https://i.imgur.com/9ULxytk.png

<!--TOC_START-->
## Contents
- [Overview](#overview)
	- [Application Load Balancing](#application-load-balancing)
	- [Network Load Balancing](#network-load-balancing)
	- [Classic Load Balancing](#classic-load-balancing)
- [Tutorial](#tutorial)
- [Exercises](#exercises)

<!--TOC_END-->
## Overview

> Load balancers are used to distribute incoming traffic between the target Virtual Machines. Amazon takes it further by allowing you to distribute between containers, IP Addresses, and Lambda functions. 

Reviewer: ```suggestion
Load balancers are used to distribute incoming traffic between the target Virtual Machines. 
Amazon takes it further by allowing you to distribute between containers, IP Addresses, and Lambda functions. 
```

> Elastic Load Balancing offers 3 types of load balacing:

Reviewer: ```suggestion
Elastic Load Balancing service offers 3 types of load balancing:
```

### Application Load Balancing

> This type of load balancing is best suited for load balacing HTTP and HTTPS traffic and offer advance routing. This means that if you have multiple microservices using REST API, then **Application Load balancer** can distribute the traffic according to the REST API call.

Reviewer: ```suggestion
This type of load balancing is best suited for load balancing HTTP and HTTPS traffic and offer advance routing. 
This means that if you have multiple micro-services using REST API, then **Application Load balancer** can distribute the traffic according to the REST API call.
```

### Network Load Balancing

> This type of load balancing is best suited for load balancing **Transmission Control Protocol (TCP)**, **User Diagram Protocol (UDP)**, and **Transport Layer Security (TLS)** where extreme performance is **REQUIRED**. This routes targets within an **Amazon Virtual Private Cloud (VPC)** and is capable of handling **millions** of requests per second, while maintaining low latencies. Must remember that this is the most expensive of all the load balancers offered by Amazon.

Reviewer: ```suggestion
This type of load balancing is best suited for load balancing **Transmission Control Protocol (TCP)**, **User Diagram Protocol (UDP)**, and **Transport Layer Security (TLS)** where extreme performance is **REQUIRED**. 
This routes targets within an **Amazon Virtual Private Cloud (VPC)** and is capable of handling **millions** of requests per second, while maintaining low latencies. 
Must remember that this is the most expensive of all the load balancers offered by Amazon.
```

### Classic Load Balancing

> This is a basic load balancer, providing load balancing between multiple EC2 instances. This is best suited for applications developed to work with an EC2 network. So having multple EC2 instances running the same application and the traffic is distributed amongst those EC2 instances.

Reviewer: ```suggestion
This is a basic load balancer, providing load balancing between multiple EC2 instances. 
This is best suited for applications developed to work with an EC2 network. 
So having multiple EC2 instances running the same application and the traffic is distributed between those EC2 instances.
```

## Tutorial

1. Navigate to the AWS Console and sign in [here][aws-signin]

2. Search for EC2 under the services dropdown menu, and click on EC2, shown in the screenshot.

![ec2-search][ec2-search]

> 3. Based on the ec2 topic where you created an Instance. This time create 2 instances and make sure they meet the following requirement:

Reviewer: ```suggestion
3. Based on the ec2 topic where you created an Instance. 
This time create 2 instances and make sure they meet the following requirement:
```

    1. Ubuntu 18.04 LTS
    2. t2.micro
    3. 2 instances

Leave the rest as it's default setting. Ensure you have the **private key** to SSH onto the server.

> 4. You will need to add tags to each of your server to help you remember by name instead of ip address. Check the server you want to provide a name to, navigate to the **Tags** tab, then click the **Add/Edit Tags**. And fill the **Key** with **Name** and **Value** with **Server1**. Repeat this for the second server.

Reviewer: ```suggestion
4. You will need to add tags to each of your server to help you remember by name instead of ip address.
Check the server you want to provide a name to, navigate to the **Tags** tab, then click the **Add/Edit Tags**. 
And fill the **Key** with **Name** and **Value** with **Server1**. Repeat this for the second server.
```

![add-tag][add-tag]

5. SSH into Server 1 and run the following commands:

```bash
sudo apt update -y

sudo apt install nginx -y

```

> 6. Since NGINX is running, you should be able to copy and paste in the ip address of Server 1 and should return the NGINX homepage. However, the security group is not allowing us to hit port 80, the default port for HTTP request, so we will need to open that port up for us to see the NGINX homepage. 

Reviewer: ```suggestion
6. Since NGINX is running, you should be able to copy and paste in the ip address of Server 1 and should return the NGINX homepage. 
However, the security group is not allowing us to hit port 80, the default port for HTTP request, so we will need to open that port up for us to see the NGINX homepage. 
```

Navigate to the description of your **Server 1** and clikc on the name of your security group, the screenshot provided shows that my security group is called **launch-wizard-2**. 

![finding-security-group][finding-security-group]

> 7. You will then need to add a new inbound rule which allows other IP Addresses to communicate with port 80. This can be done by, navigating to the **inbound** tab, clicking on the **Edit** button, then clicking on **Add Rule** button and add in the rule shown on the screenshot.

Reviewer: ```suggestion
7. You will then need to add a new inbound rule which allows other IP Addresses to communicate with port 80. 
This can be done by, navigating to the **inbound** tab, clicking on the **Edit** button, then clicking on **Add Rule** button and add in the rule shown on the screenshot.
```

> Since both servers are using the same security group you only need to make this change once and both servers will be updated. Now, copy and paste the ip address of **Server 1** onto the browser and you should see the NGINX homepage.

Reviewer: ```suggestion
Since both servers are using the same security group you only need to make this change once and both servers will be updated. 
Now, copy and paste the ip address of **Server 1** onto the browser and you should see the NGINX homepage.
```

![adding-inbound-rule][adding-inbound-rule]

> 8. You will need to modify the **index.html** which is showing the NGINX homepage so that you can identify which server is going to be returned by the Load balancer. Run the following commands:

Reviewer: ```suggestion
8. You will need to modify the **index.html** which is showing the NGINX homepage so that you can identify which server is going to be returned by the Load balancer. 
Run the following commands:
```

```bash

sudo su

echo > /var/www/html/index.nginx-debian.html

nano /var/www/html/index.nginx-debian.html

```
Copy the following into the .html file

```bash

<html>
<head>
<h1>Hello Community!</h1>
<h2>This is our First Server</h2></head>
<body bgcolor="#98fb98">
</body>
</html>

```

> Then you will need to reload your NGINX server to make sure the changes can be seen. Run the following command to reload NGINX:

Reviewer: ```suggestion
Then you will need to reload your NGINX server to make sure the changes can be seen. 
Run the following command to reload NGINX:
```

```bash

sudo service nginx reload

```

> 9. You will then need to repeat steps 5 and 8 for the second server. This time making sure the HTML code you are pasting in is altered so that it says **This is our Second Server** and use a different colour. Here is an example of a different colour: 5DBCD2

Reviewer: ```suggestion
9. You will then need to repeat steps 5 and 8 for the second server. 
This time making sure the HTML code you are pasting in is altered so that it says **This is our Second Server** and use a different colour. 
Here is an example of a different colour: 5DBCD2
```

Now we are ready to create a load balancer.

10. Navigate to the Load balancer dashboard [here][load-balancer-dashboard]

11. Click the **Create Load Balancer** button

![create-load-balancer-button][create-load-balancer-button]

12. You're going to create an **Application Load balancer**, click on the **Create** button in the application load balancer section

![create-application-load-balancer][create-application-load-balancer]

> 13. Fill in the fields according to the screen shots provided, then click **Next: Configure Security Group**. You might get a notification in yellow telling you about how if you want to use HTTPS you need to use port 443 instead. Just ignore this message and continue to configure security group.

Reviewer: ```suggestion
13. Fill in the fields according to the screen shots provided, then click **Next: Configure Security Group**.
You might get a notification in yellow telling you about how if you want to use HTTPS you need to use port 443 instead. 
Just ignore this message and continue to configure security group.
```

![load-balancer-configuration][load-balancer-configuration]

> 14. You will need to create a **New** security group. This is specific to your load balancer. By default, the **Existing Security Groups** will be checked, so change this to **Create a new Security Group**. This will then by default have an inbound rule which will allow HTTP request, which in this case is what we want.Leave everything in its default settings and proceed to **Configuring Route**.

Reviewer: ```suggestion
14. You will need to create a **New** security group. This is specific to your load balancer. By default, the **Existing Security Groups** will be checked, so change this to **Create a new Security Group**. 
This will then by default have an inbound rule which will allow HTTP request, which in this case is what we want.Leave everything in its default settings and proceed to **Configuring Route**.
```

> 15. You will be configuring where the user should go to when traffic is sent to this load balancer. Leave everything in its default setting and only give the group it's name. Proceed to **Register Targets** 

Reviewer: ```suggestion
15. You will be configuring where the user should go to when traffic is sent to this load balancer. Leave everything in its default setting and only give the group it's name. 
Proceed to **Register Targets** 
```

![load-balancer-routing-config][load-balancer-routing-config]

> 16. Here you will select the Instances you want the load balancer to balance the traffic between. Select the 2 Instances which you created earlier and click **Add to registered** and proceed to **Review**

Reviewer: ```suggestion
16. Here you will select the Instances you want the load balancer to balance the traffic between. 
Select the 2 Instances which you created earlier and click **Add to registered** and proceed to **Review**
```

![load-balancer-add-instances][load-balancer-add-instances]

17. If you're happy with the configuration, click **Create**

> 18. The load balancer is being created and it might take some time before the load balancer starts to work. Once you think it is ready, copy the DNS onto a browser and you should see the page which you created for one of your server. If you keep refreshing your page, you will notice that it changes to the other server.

Reviewer: ```suggestion
18. The load balancer is being created and it might take some time before the load balancer starts to work. 
Once you think it is ready, copy the DNS onto a browser and you should see the page which you created for one of your server. 
If you keep refreshing your page, you will notice that it changes to the other server.
```

Navigate to the description of your **Server 1** and clikc on the name of your security group, the screenshot provided shows that my security group is called **launch-wizard-2**. 

![finding-security-group][finding-security-group]

7. You will then need to add a new inbound rule which allows other IP Addresses to communicate with port 80. 
This can be done by, navigating to the **inbound** tab, clicking on the **Edit** button, then clicking on **Add Rule** button and add in the rule shown on the screenshot.

Since both servers are using the same security group you only need to make this change once and both servers will be updated. 
Now, copy and paste the ip address of **Server 1** onto the browser and you should see the NGINX homepage.

![adding-inbound-rule][adding-inbound-rule]

8. You will need to modify the **index.html** which is showing the NGINX homepage so that you can identify which server is going to be returned by the Load balancer. 
Run the following commands:

```bash

sudo su

echo > /var/www/html/index.nginx-debian.html

nano /var/www/html/index.nginx-debian.html

```

> Copy the following into the .html file

Reviewer: ```suggestion
Copy the following into the .html file, use SHIFT+INSERT to paste it and keep the formatting
```

```bash

<html>
<head>
<h1>Hello Community!</h1>
<h2>This is our First Server</h2></head>
<body bgcolor="#98fb98">
</body>
</html>

```

Then you will need to reload your NGINX server to make sure the changes can be seen. 
Run the following command to reload NGINX:

```bash

sudo service nginx reload

```

9. You will then need to repeat steps 5 and 8 for the second server. 
This time making sure the HTML code you are pasting in is altered so that it says **This is our Second Server** and use a different colour. 
Here is an example of a different colour: 5DBCD2

Now we are ready to create a load balancer.

> 10. Navigate to the Load balancer dashboard [here][load-balancer-dashboard]

Reviewer: ```suggestion
10. Navigate to the Load balancer dashboard using the following link [here][load-balancer-dashboard], or find it in the left navigation menu
```

This will then by default have an inbound rule which will allow HTTP request, which in this case is what we want.Leave everything in its default settings and proceed to **Configuring Route**.

15. You will be configuring where the user should go to when traffic is sent to this load balancer. Leave everything in its default setting and only give the group it's name. 
Proceed to **Register Targets** 

![load-balancer-routing-config][load-balancer-routing-config]

16. Here you will select the Instances you want the load balancer to balance the traffic between. 
Select the 2 Instances which you created earlier and click **Add to registered** and proceed to **Review**

![load-balancer-add-instances][load-balancer-add-instances]

17. If you're happy with the configuration, click **Create**

18. The load balancer is being created and it might take some time before the load balancer starts to work. 
Once you think it is ready, copy the DNS onto a browser and you should see the page which you created for one of your server. 
If you keep refreshing your page, you will notice that it changes to the other server.

![load-balancer-DNS][load-balancer-DNS]

> 

Reviewer: add a section listing out the resources that were created in this tutorial so that if someone wants to clean up after the exercise you don't need to go through the exercise again to know what you've created during it

Author: I have added a list of resources used from this tutorial. This is under the section, 'Resources'.

> 

Reviewer: add a section of now how to test it and what to expect, like where do you get the ip to test it, and what changes on the page to indicate that it's working

Author: I have provided them a set of screenshots under the section, 'Test'. As I did not make them get the IP address from each server, all it was a different static website running on the different servers served by NGINX. If they followed the tutorial they should see the expected outputs shown in the screenshots.

![load-balancer-configuration][load-balancer-configuration]

14. You will need to create a **New** security group. This is specific to your load balancer. By default, the **Existing Security Groups** will be checked, so change this to **Create a new Security Group**. 
This will then by default have an inbound rule which will allow HTTP request, which in this case is what we want.Leave everything in its default settings and proceed to **Configuring Route**.

15. You will be configuring where the user should go to when traffic is sent to this load balancer. Leave everything in its default setting and only give the group it's name. 
Proceed to **Register Targets** 

![load-balancer-routing-config][load-balancer-routing-config]

16. Here you will select the Instances you want the load balancer to balance the traffic between. 
Select the 2 Instances which you created earlier and click **Add to registered** and proceed to **Review**

![load-balancer-add-instances][load-balancer-add-instances]

17. If you're happy with the configuration, click **Create**

18. The load balancer is being created and it might take some time before the load balancer starts to work. 

> Once you think it is ready, copy the DNS onto a browser and you should see the page which you created for one of your server. 

Reviewer: ```suggestion
Once you think it is ready, copy the DNS name onto a browser and you should see the page which you created for one of your server. 
```

If you keep refreshing your page, you will notice that it changes to the other server.

![load-balancer-DNS][load-balancer-DNS]

## Test

### Overview
When you load the DNS for the load balancer, you should be expecting to see the static website returned from each EC2 instances. If you refresh the page your load balancer will return values from 2 different EC2 instances and you should be expecting to see the followings if you keep refreshing the page:

![load-balancer-test-1][load-balancer-test-1]

![load-balancer-test-2][load-balancer-test-2]

## Resources

### Overview
These are the list of resources used for this topic:

1. 2 EC2 Instances
2. Elastic Load Balancer

> 

Reviewer: ```suggestion
3. Security Group
```
